### PR TITLE
Add docfx and documentation style guide

### LINKED ISSUE #102: Create articles on docfx and documentation style guide
### Motivation
The current README which hosts instructions on how to create documentation with docfx is out of date and a little short on detail. It would also be good to standardize best practices for documentation, which currently is spread mainly through discussions with package developers/documentation writers.  A previous attempt was made at #82, and as part of the feedback for that PR, it was brought up that it might be better to package this as an article. This would also serve as a good base for discussion during the documentation session for the upcoming [Bonsai Developer Conference](https://conference.bonsai-rx.org/2024/).

### Proposed solution
While I was working on it, I found that it was getting too long and thought there was enough content to split it into two articles. Thus I propose making the following changes:

- **Documentation with docfx**- This would cover repository and docs folder organization, installing docfx and docfx configuration, publishing to github pages, and troubleshooting tips

- **Documentation style guide** - This would cover article organization, constructing the table of contents, using individual operator articles and the overwrite function, bonsai workflow image export and containers, standard formatting for operators and other common Bonsai content.

- **README.md** - Trim it to just the basics for people to get started with contribution, with links to the articles on docfx and documentation style guide.

## Additional issues to consider
- Article Location - currently the articles does not fit nicely into the existing article categories, but `Extending Bonsai` is the closest since there is also a `Create a Package` article. I think creating a new section for `Custom Package Development` or even a `Bonsai Developers` subsection of the website would be a better fit (as it is not technically a manual article for normal Bonsai Users). There might not be enough content to go with the later option, but I think the package development articles also need to be expanded (unfortunately I have zero experience with that)

### PR DESCRIPTION
**Pull request by banchan86**
Followup to #82, resolves #102.
Accidentally closed the original PR as I was shifting things around but have reproduced the comments below.
This PR introduces two new articles a "Documentation with docfx" and "Documentation style guide"

Review breakdown:
- "We might want to consider mentioning here how to build the solution in order to export the example workflows, and in that case also consider how to automate the msbuild part into the powershell module containing the export image functions."

  - I have included sections on creating a local environment, setting up `build.ps1` in the documentation docfx article, and added more detail to the Bonsai workflow section of the documentation style guide. Its a little over the place, and I think I placed things where they would logically belong, but happy to revise. Im not sure I will be able to carry out 2nd part of that comment.

- "We should maybe include these as part of docfx-tools or some other template repo."

  - I initially raised a [PR](https://github.com/bonsai-rx/docfx-tools/pull/2) to include them as part of the docfx-tools submodule but i realized that maybe wasn't the best solution as users would still need to modify and move them. I am more in favor of hosting them on the [docfx-assets](https://github.com/bonsai-rx/docfx-assets) repository for users to download but not import as a submodule and have raised a [PR](https://github.com/bonsai-rx/docfx-assets/pull/2) to add them. I have also updated the documentation articles with that strategy in mind.

- "We might want to consider marking the Tutorials section as optional for other package documentation, to make things simpler as a common reference." 

  - I have marked this more explicitly as being optional in the top section of the documentation style guide

- "Review format of relative / local URLs to make sure we have a solution that works for both local and remote builds."

  - So I went to research this and from what I can tell its safer to use `../` rather than `~/` and given that our scripts/code use `../` for navigation, I thought it would be best to standardize on using `./` and `../`. I have tested this on docfx articles and it works and I have changed instances of `~/` as well as backward slashes in the documentation articles to reduce ambiguity.

- "It might be worth mentioning as a note the importance of matching the publishing URL with the PackageProjectUrl attribute specified in the .csproj files. This is where the Bonsai editor will look for the xrefmap.yml file to retrieve URLs for operators." 
 
  - Have added a note to the publishing to github pages section.
 
Other things that I have changed or moved around since the last time:

- Decided to use directory trees to better represent file locations and simplify the text in those sections
- Updated the section that deals with the local bonsai environment given that I now have a better understanding of how that works.
- Restructured the content in a way I think makes more sense (that lead to the creation of the repository organization, doc organization and version control cleanup sections)
- Expanded the text so that it reads less like a README and more like an article with proper explaination for why we do certain things, but I may have taken it too far and made it too wordy :)

Online preview of this PR and PRs #97 and #94 : https://banchan86.github.io/bonsai_docs_test/articles/documentation-docfx.html

### CODE REVIEW

File: articles/documentation-style-guide.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+```markdown
+Other 
+[!include[Title](../apidoc/Bonsai_Sleap_PredictPoses.md)]
+```
+> [!NOTE]  
+> The title is optional.
+
+### Reference section
+
+As mentioned the `Reference` section of the website is auto-generated from [XML documentation comments](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags) in the code. A unique feature we have integrated into the Bonsai Editor is that when users when users right click on individual operators they will be shown the `Reference` page for that operator. For many users, this might be the most obvious entry point to the package documentation. Thus it is essential that developers ensure that:
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:97]:
```suggestion
As mentioned the `Reference` section of the website is auto-generated from [XML documentation comments](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/recommended-tags) in the code. We have integrated into the Bonsai Editor the ability to right-click on individual operators and select `View Help`. This will bring up and load the `Reference` page for that operator in the editor itself. For most people, this might be the most obvious entry point to the package documentation. Thus it is essential that developers ensure that:
```

File: articles/documentation-style-guide.md [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+1) XML documentation comments are clear, simple and helpful.
+2) Where appropriate, supplement additional information (basic workflow, images, long-form content) with an individual operator article using the `overwrite` function detailed in the previous section.
+
+At present, we understand that the default template for the `Reference` page is a little too technically oriented, and somewhat confusing for Bonsai users to navigate. We are in the midst of revamping the template to cater to a broader audience and will update this section when it becomes availiable.
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:102]:
```suggestion
At present, we understand that the default template for the `Reference` page is a little too technical, and somewhat confusing to navigate. We are in the midst of revamping the template to cater to a broader audience and will update this section when it becomes available.
```

File: articles/documentation-style-guide.md [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+### Tutorials section
+
+For the tutorials section, we suggest creating individual pages for various applications where the package is likely to be used. For each application, guide learners through the workflow step by step, organizing it into separate exercises with clear objectives and visible results. This structure makes the content more accessible and manageable for learners, especially for complex workflows.
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:106]:
```suggestion
For the tutorials section, we suggest creating individual pages for various applications where the package is likely to be used. For each application, try to guide learners through the workflow step by step, organizing it into separate exercises with clear objectives and visible results. This structure makes the content more accessible and manageable, especially for complex workflows.
```

File: articles/documentation-style-guide.md [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+#### Submodule
+For packages with extensive tutorials, multimedia and other large files, a separate repository can be created and imported as a submodule.
+
+A tutorial submodule can be added with the following command in the `docs` directory:
+
+```powershell
+git submodule add https://github.com/bonsai-rx/machinelearning-examples
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:114]:
```suggestion
git submodule add https://github.com/bonsai-rx/package-examples
```

Using a non-existing repo name here to make sure people are forced to change the command line.

File: articles/documentation-style-guide.md [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+In addition, the `docfx.json` file needs to be modified to import the relevant resources. For an example of how to setup a Tutorial submodule, refer to the [Bonsai Machine Learning Package](https://github.com/bonsai-rx/machinelearning) package and its submodule [Machine Learning Examples](https://github.com/bonsai-rx/machinelearning-examples).
+
+## Article formatting guide 
+
+In docfx, articles are written in [Markdown](https://dotnet.github.io/docfx/docs/markdown.html?tabs=linux%2Cdotnet) and rendered with the [Markdig](https://github.com/xoofx/markdig) parsing engine. Refer to these links for more information on how to format content in markdown.
+
+When writing articles, please follow the [MSDN writing tips](https://learn.microsoft.com/en-us/style-guide/global-communications/writing-tips). In particular:
+
+- Keep article and section titles short and succinct so that the table of contents that appears on the left and right sidebar are easier to read (and also to assist in machine translation)
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:125]:
```suggestion
- Keep article and section titles short and succinct so that the table of contents that appears on the left and right sidebar are easier to read (and also to assist in machine translation).
```

File: articles/documentation-style-guide.md [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+- Reuse operator names, properties, and descriptions in the articles and titles (do not use synonyms) so that readers may more easily follow and refer to them.
+- Use the imperative style i.e. "Link hardware triggers" rather than "Linking hardware triggers".
+
+### Linking content
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:129]:
```suggestion
### Content links 
```

File: articles/documentation-style-guide.md [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+For content in a different folder, use `../` to navigate up to the parent folder, and then specify the folder with the content like this:
+
+```markdown
+../images/bonsai-screenshot.svg
+```
+
+### Cross-references and highlights
+
+When working on an article, first check [the main documentation](https://bonsai-rx.org/docs/) to see what written material might already exist that could be linked if it would be helpful and to avoid duplication. Link resources like websites for external libraries for users to explore for more information.
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:148]:
```suggestion
When working on an article, first check [the main documentation](https://bonsai-rx.org/docs/) to see what written material might already exist that could be linked if it would be helpful and to avoid duplication. Link resources like websites for external libraries which may be useful to explore for more information.
```

File: articles/documentation-style-guide.md [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+- Set the `Pin` property of the [`DigitalOutput`](xref:Bonsai.Arduino.DigitalOutput) operator to 13.
+- Configure the `PortName` property.
+- Run the workflow and change the `Value` property of the [`Boolean`](xref:Bonsai.Expressions.BooleanProperty) operator.
+```
+
+You can also link to operators in other packages if their [xrefmap](./documentation-docfx.md#configuring-docfx) is included in `docfx.json`.
+
+### Bonsai workflows
+
+To include and/or reference an example workflow in an article of the documentation, first create the example workflow in a Bonsai workflow editor and save the workflow as `articleFileName-workflowName`.
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:199]:
```suggestion
To include and/or reference an example workflow in an article of the documentation, first create the example workflow in a Bonsai workflow editor and save the workflow as `articleFileName-workflowName.bonsai`. 
```

File: articles/documentation-style-guide.md [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+Assuming you want to include `CustomPulseTrain-SendCustomWaveform.bonsai`: 
+
+```markdown
+:::workflow
+![Send Custom Waveform](../workflows/CustomPulseTrain-SendCustomWaveform.bonsai)
+:::
+```
+
+Workflow images are automatically exported as SVGs by the [docfx-tools](https://github.com/bonsai-rx/docfx-tools) submodule and requires the [build.ps1](./documentation-docfx.md#docfx-folder-organization) file and a [local bonsai environment](./documentation-docfx.md#repository-organization).
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:212]:
```suggestion
Workflow images are automatically exported as SVG files by the [docfx-tools](https://github.com/bonsai-rx/docfx-tools) submodule and requires the [build.ps1](./documentation-docfx.md#docfx-folder-organization) file and a [local bonsai environment](./documentation-docfx.md#repository-organization). 
```

**Review by glopesdev — COMMENTED** [review]
Thanks for putting this together @banchan86, it's a really great start. I did a full first pass over the two guides, this is definitely something we want to have merged soon.

The only thing I am still struggling with is how to move with general infrastructure decisions such as `docfx-assets`, the centralization or decentralization of package documentation, package documentation index and discoverability, etc.

I will try to make some definite progress on this so we can potentially review this soon, but in the meantime I'm leaving some general feedback that can be worked on independently.

If we can't resolve some of these fundamental questions soon we may want to merge this anyway and iterate later, since this is clearly much better than what we have now (which is either nothing or obsolete content!).

File: README.md [hunk 10]
```diff
@@ -4,70 +4,26 @@
...
-6. Once a piece of documentation has been polished, and at least two community members agree that it is ready for a final review, create a pull request to merge with the "main" branch of this repo. At this point, @glopesdev will conduct the final review and quality control check. If the contribution passes this final step, the PR to merge with "main" will be approved and the contribution will be published on the [publicly accessible Bonsai documentation website](https://bonsai-rx.org/docs-wip/).
+We welcome all contributions! For more information please refer to our [Contribution Guide](https://bonsai-rx.org/contribute/).
 
-## Contributor Style Guide 
+# Quick start guide for contributing to documentation
 
-### Why use the "develop" branch?
+These instructions apply to repos that already have a docfx website created.
```
[the remark this answers is not in the document]
Comment (banchan86) [reply README.md:16]: So I have struggled with this in the beginning as well, I think that actually "docfx" is the way that they have now intended it to be, but I think they are still having trouble with maintaining it consistently as you can still find instances of "Docfx" and even the older "DocFX" peppered throughout their docs. For me I relied on several signposts from https://dotnet.github.io/docfx/:

- Their footer uses "Made with docfx"
- The docfx website logo uses "docfx"
- The landing page (index.md -> quick start guide) updated 3 weeks ago uses "docfx" and only capitalizes "Docfx" when it is in the beginning of the sentence.

As such, I think I would prefer to keep it the way it is.

File: articles/documentation-docfx.md [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+Effective online documentation plays a crucial role in driving the adoption and usage of custom packages. For packages published by the Bonsai Foundation, we use [docfx](https://dotnet.github.io/docfx/index.html), a static site generator for .NET projects, to automatically generate and publish documentation online. As we have integrated and extended `docfx` to facilitate key Bonsai features, we encourage package developers to use `docfx` for consistency and support within the Bonsai Ecosystem. 
+
+The following article will go into how to create a `docfx` website for your package. For documentation style guidelines, check out the [Documentation Style Guide](./documentation-style-guide.md).
+
+> [!TIP]
+> We recommend using a code editor with an integrated terminal like [Visual Studio Code](https://code.visualstudio.com/) to follow along with these instructions.
+
+## Repository organization
+For effective long term maintenance of documentation across different repositories, package repositories should be organised according to this standardized folder structure before setting up `docfx`:
```
[the remark this answers is not in the document]
Comment (banchan86) [reply articles/documentation-docfx.md:11]: Just as a stylistic point for me to take note of the future, we are using American English? I will try to be more careful about it in the future (I have used both and mix it up sometimes)

File: README.md [hunk 12]
```diff
@@ -4,70 +4,26 @@
...
-- Run the workflow and change the [`Value`](xref:Bonsai.Expressions.BooleanProperty.Value) property of the [**`Boolean`**](xref:Bonsai.Expressions.BooleanProperty) operator.
-- **Optional:** Use your mouse to control the LED! Replace the [**`Boolean`**](xref:Bonsai.Expressions.BooleanProperty) operator by a `MouseMove` source (hint: use `GreaterThan`, `LessThan`, or equivalent operators to connect one of the mouse axis to [**`DigitalOutput`**](xref:Bonsai.Arduino.DigitalOutput).
+1. Fork the repository you want to contribute documentation to. 
+2. Ensure that you are working off the `main` branch in your fork and create a descriptively named branch for each article/issue that you want to work on.
+3. Download and install dotnet (https://dotnet.microsoft.com/en-us/download)
+4. Restore `docfx` by running
+```powershell
+dotnet tool restore
 ```
+5. In a Windows Powershell, use the command `dotnet docfx docfx.json --serve` to generate a local preview of the documentation website as you are making changes.
```
[the remark this answers is not in the document]
Comment (banchan86) [reply README.md]: So I was debating what to keep from the original README and how to structure the content to avoid duplication.

I think we should keep the README short and simple, as people who are looking to make contributions to documentation might not need to know the ins and outs of docfx as documented in the `documentation-docfx.md` or even maybe the style guidelines in `documentation-style-guide.md` (if they just follow the formatting of existing articles). It looks less intimidating, contains just the minimum necessary information and lowers the barrier to entry for contributions. And then we have links to those who want to explore more, need more guidance or face problems with docfx. 

I think the stepping into `docs` folder is a good point though and I will add it because the main docs website is structured differently.

File: README.md [hunk 13]
```diff
@@ -4,70 +4,26 @@
...
-Add the `.svg` file to the **images** folder in this repo, and add the `.bonsai` file to the **workflows** folder in this repo. In the text of the article that includes/references this example workflow, add the `.svg` file as an image and link that image to the `.bonsai` file. 
-
-**Example:**
-
-Assuming you want to include `acquisition-example.bonsai`: 
-
-```markdown
-[![Example Workflow](~/images/acquisition-example.svg)](~/workflows/acquisition-example.bonsai)
-```
+For more info, check out the [Documentation with docfx](https://bonsai-rx.org/docs/articles/documentation-docfx.html) and [Documentation Style Guide](https://bonsai-rx.org/docs/articles/documentation-style-guide.html) articles.
```
[the remark this answers is not in the document]
Comment (banchan86) [reply README.md:29]: See the comment I left above, for that reason I think I would resist folding more content into the README but let me know what you think

File: README.md [hunk 14]
```diff
@@ -4,70 +4,26 @@
...
-5. When you are ready to have your contribution reviewed, commit your edits to the approriate branch of your fork of the repo and create a PR to merge that branch with the "develop" branch of this original repo. Because the "develop" branch was created explicitly to aid collaboration between contributors, pull requests to this branch will be accepted and merged rapidly (at least once a week). 
-6. Once a piece of documentation has been polished, and at least two community members agree that it is ready for a final review, create a pull request to merge with the "main" branch of this repo. At this point, @glopesdev will conduct the final review and quality control check. If the contribution passes this final step, the PR to merge with "main" will be approved and the contribution will be published on the [publicly accessible Bonsai documentation website](https://bonsai-rx.org/docs-wip/).
+We welcome all contributions! For more information please refer to our [Contribution Guide](https://bonsai-rx.org/contribute/).
```
[the remark this answers is not in the document]
Comment (banchan86) [reply README.md]: Good point, I think it might be confusing to people, I have reused the contribution section from the main bonsai repo README.

File: articles/documentation-docfx.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+To update the submodule to the latest commit:
+```powershell
+git submodule update --remote
+```
+
+- `template/` - this custom CSS template patches in docfx-tools and adds a github link to the top right of the navigation bar. Download this folder from the [docfx-assets](https://github.com/bonsai-rx/docfx-assets) repository and amend the github link in `main.css` to your repository.
+
+- `build.ps1` - this custom script exports Bonsai workflow images using `docfx-tools` and then calls the docfx build process. Download the file from the [docfx-assets](https://github.com/bonsai-rx/docfx-assets) repository and amend the line that specifies the package name and source location.
+
+- `logo.svg` and `favicon.ico` - for official Bonsai packages these can be downloaded from the [docfx-assets](https://github.com/bonsai-rx/docfx-assets) repository.
```
[the remark this answers is not in the document]
Comment (banchan86) [reply articles/documentation-docfx.md]: I did start creating some python scripts to automate some of the docfx setup (see https://github.com/bonsai-rx/docfx-tools/pull/11 for a script to configure docfx.json) but now that you mention it, Bonsai.Templates might be a better place to have a documentation template. I guess this will also solve the issue of where to put the assets. Is there a GitHub repository for it to raise an issue so we can keep track of it?

**Comment by banchan86** [comment]
> Thanks for putting this together @banchan86, it's a really great start. I did a full first pass over the two guides, this is definitely something we want to have merged soon.
> 
> The only thing I am still struggling with is how to move with general infrastructure decisions such as `docfx-assets`, the centralization or decentralization of package documentation, package documentation index and discoverability, etc.
> 
> I will try to make some definite progress on this so we can potentially review this soon, but in the meantime I'm leaving some general feedback that can be worked on independently.
> 
> If we can't resolve some of these fundamental questions soon we may want to merge this anyway and iterate later, since this is clearly much better than what we have now (which is either nothing or obsolete content!).

Sound good, thanks for reviewing it!

File: articles/documentation-docfx.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md]: Yes, for code I tend to be biased towards american english (which still dominates use in code naming conventions, etc). However I admit not being entirely consistent myself!

File: articles/documentation-docfx.md [hunk 16]
```diff
@@ -2,33 +2,33 @@
...
+├── .github/workflows/       # GitHub Actions Folder (see note below) 
 │   └── docs.yml
 ├── docs/                    # Folder to setup docfx
 └── src/
     └── Bonsai.PackageName/  # Project files here
 ... 
 ```
 
-- `.bonsai/` - A local Bonsai [environment](https://bonsai-rx.org/docs/articles/environments.html) is required for a script that exports Bonsai workflow images. To create one run these following commands:
+- `.bonsai/` - A local Bonsai [environment](https://bonsai-rx.org/docs/articles/environments.html) is required to run the script that exports Bonsai workflow images. To create one, run the following commands:
```
Comment (glopesdev) [inline articles/documentation-docfx.md:24]:
```suggestion
- `.bonsai/` - A local Bonsai [environment](./environments.md) is required to run the script that exports Bonsai workflow images. To create one, run the following commands:
```

Noticing now this and a few other links could be replaced with local markdown references, useful when debugging local docs preview.

File: articles/documentation-docfx.md [hunk 17]
```diff
@@ -252,9 +252,9 @@ To keep your online Github repository clean, you can use .gitignore files to ign
 
 ## Testing unpublished packages
 
-To write documentation for [new packages or releases](https://bonsai-rx.org/docs/articles/create-package.html) that have not been published to the community, test them in Visual Studio. 
+To write documentation for [new packages or releases](https://bonsai-rx.org/docs/articles/create-package.html) that have not yet been published to the community, it is good practice to start by writing test workflows in Visual Studio.
```
Comment (glopesdev) [inline articles/documentation-docfx.md:255]:
```suggestion
To write documentation for [new packages or releases](./create-package.md) that have not yet been published to the community, it is good practice to start by writing test workflows in Visual Studio.
```

File: articles/documentation-docfx.md [hunk 18]
```diff
@@ -252,9 +252,9 @@ To keep your online Github repository clean, you can use .gitignore files to ign
...
 
-1) Install [Visual Studio](https://www.visualstudio.com/) (the community edition can be installed for free).
+1) Install [Visual Studio](https://www.visualstudio.com/) (the community edition can be installed for free for open-source projects).
 2) Install Bonsai VS Extensions. Assuming Bonsai is already installed, from the Windows Start Menu, search for the "Install Bonsai VS Extensions" shortcut and run it.
 3) In Visual Studio, open `src/PackageName.sln` in the repository.
 4) Press F5 to open the Bonsai editor with the new package added.
```
Comment (glopesdev) [inline articles/documentation-docfx.md:260]:
```suggestion
4) Press F5 to open the Bonsai editor with the new package added.
```

Remove whitespace at end of line

File: articles/create-package.md
Comment (glopesdev) [inline articles/create-package.md:1]: Line 116: replace absolute URL: `https://bonsai-rx.org/docs/articles/packages.html#configure-package-sources` with `./packages.md#configure-package-sources`

File: articles/documentation-style-guide.md [hunk 19]
```diff
@@ -145,7 +145,7 @@ For content in a different folder, use `../` to navigate up to the parent folder
 
 ### Cross-references and highlights
 
-When working on an article, first check [the main documentation](https://bonsai-rx.org/docs/) to see what written material might already exist that could be linked if it would be helpful and to avoid duplication. Link resources like websites for external libraries for users to explore for more information.
+When working on an article, first check [the main documentation](https://bonsai-rx.org/docs/) to see what written material might already exist that could be linked if it would be helpful and to avoid duplication. Link resources like websites for external libraries which may be useful to explore for more information.
```
Comment (glopesdev) [inline articles/documentation-style-guide.md:148]:
```suggestion
When working on an article, first check [the main documentation](../index.md) to see what written material might already exist that could be linked if it would be helpful and to avoid duplication. Link resources like websites for external libraries which may be useful to explore for more information.
```

**Review by glopesdev — COMMENTED** [review]
@banchan86 thanks, all looks good to me, I just noticed a few opportunities to use relative instead of absolute links and after that it should be good to go.

Other than the files in this PR, the only other absolute link I noticed in a cursory look through the entire docs is Line 181 of `closed-loop.md` where `[property mapping operators](http://bonsai-rx.org/docs/property-mapping){:target="\_blank"}` could be replaced with `[property mapping operators](../articles/property-mapping.md).`.

We can also do that one on a separate PR but since it is just the one off maybe there is no harm in adding it here.

**Comment by glopesdev** [comment]
@banchan86 Thanks for the edits, almost there, I have added a final round of suggestions throughout in a final commit directly to the PR branch, let me know if these sound good to you. There are a few things which might still change in the docs short-term but I think I'm happy to push this version and then iterate as we improve.

**Comment by banchan86** [comment]
Looks good to me! Before pushing it though, have we decided what to do with the documentation assets? As the articles point to some files that don't currently exist yet in the `docfx-assets` repo unless https://github.com/bonsai-rx/docfx-assets/pull/2 is merged in as well.

**Comment by glopesdev** [comment]
@banchan86 good point about the assets. What I decided to do is actually embed them directly in the guide as code sections, similar to the bits about configuring `docfx.json`. Turns out most of them are really small and not very complicated, except for the GitHub Actions workflow.

For that one it actually gets even more complicated since we don't currently have yet a solid standard of how automation should be made, so I deferred with a link to an example workflow for now. We can review later when we have something more refined and stable.

If this sounds good I would be happy merging this.

**Review by banchan86 — COMMENTED** [review]
Looks good to me overall, and I agree with the approach of just embedding the code snippets for most of the files, but not sure about pointing to the main docs GitHub actions recipe.

File: articles/documentation-docfx.md [hunk 20]
```diff
@@ -28,24 +28,24 @@ dotnet new install Bonsai.Templates
 dotnet new bonsaienv
 ```
 
-- `.github/workflows/` - The `docfx` website is published to [GitHub Pages](https://pages.github.com/) using a [GitHub Actions](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions) workflow called `docs.yml`. Download this folder from the [docfx-assets](https://github.com/bonsai-rx/docfx-assets) repository and amend `Bonsai.PackageName` to point to your package source code in `docs.yml`.
+- `.github/workflows/` - The `docfx` website is often automatically published to [GitHub Pages](https://pages.github.com/) using a [GitHub Actions](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions) workflow specified in the `docs.yml` file. Unfortunately these workflows tend to be very project specific and we do not yet have a standardized pipeline. For a minimal example and inspiration you can check [the workflow for this repository](https://github.com/bonsai-rx/docs/blob/main/.github/workflows/build.yml).
```
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md:31]: That is a good point, although unfortunately most of the other repos are also using deprecated actions which I would prefer not to use as a reference going forward. I was wondering maybe the best middle ground would be to upgrade one of the old repos to try out the upgraded pipeline.

I am reviewing https://github.com/bonsai-rx/zeromq/pull/22 and will try to upgrade everything to fit correctly. Once we are done we can point to that, does that sound good?

File: articles/documentation-docfx.md [hunk 21]
```diff
@@ -250,32 +283,32 @@ To keep your online GitHub repository clean, you can use `.gitignore` files to i
 ...
 ```
 
-## Testing unpublished packages
+## Creating example workflows
```
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md:286]: Good point about the duplicate section. I think I would rather move everything to one of the sections then. Perhaps this one, since including example workflows feels really part of the core documentation infrastructure and not simply a style recommendation. Does that sound good?

Maybe I am confused about the distinction between "published" and "unpublished" packages. I was imagining that documentation would always be done targeting the source code of an existing package, and so in a way you always treat it as "unpublished" since you might be documenting a new release.

Although perhaps you were thinking of cookbook style or recipe examples and tutorials which might live outside of any package and where you just describe in detail how to use existing packages to achieve a specific goal? In that case I can see the difference, but then the entire way of setting up docfx might be different anyway and potentially deserve a separate guide? For now maybe we just explain how to document custom packages?

File: articles/documentation-docfx.md [hunk 22]
```diff
@@ -250,32 +283,32 @@ To keep your online GitHub repository clean, you can use `.gitignore` files to i
...
-1) Install [Visual Studio](https://www.visualstudio.com/) (the community edition can be installed for free for open-source projects).
-2) Install Bonsai VS Extensions. Assuming Bonsai is already installed, from the Windows Start Menu, search for the "Install Bonsai VS Extensions" shortcut and run it.
-3) In Visual Studio, open `src/PackageName.sln` in the repository.
-4) Press F5 to open the Bonsai editor with the new package added.
-5) From here, you can make Bonsai workflows and save them as per normal.
+1. Install [Visual Studio](https://www.visualstudio.com/) (the community edition can be installed for free for open-source projects).
+2. Install Bonsai VS Extensions. Assuming Bonsai is already installed, from the Windows Start Menu, search for the "Install Bonsai VS Extensions" shortcut and run it.
+3. In Visual Studio, open `src/PackageName.sln` in the repository.
+4. Press F5 to open the Bonsai editor with the new package added.
+5. From here, you can make Bonsai workflows and save them to the `workflows` folder as examples.
```
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md:294]: Good point, I will rephrase.

File: articles/documentation-style-guide.md [hunk 23]
```diff
@@ -6,14 +6,14 @@ Through our experience documenting Bonsai, we have established a set of recommen
 
 In general, we have converged on three types of articles, organized into distinct website sections, to address the needs of most people and their different learning styles.
 
-* Manual - hosts documentation that explains the basic workflow of the package or functions of the various operators.
-* Reference - hosts technical documentation for each operator, generated automatically by `docfx` from XML comments in source code or supplemented with individual operator articles.
-* Tutorials - hosts examples or tutorials for various applications. This section is optional, but valuable for more complicated applications or packages which require operators from other packages for their execution.
+* `Manual` - articles explaining the core concepts of the package and functions of its various operators.
+* `Reference` - technical documentation for each operator, generated automatically by `docfx` from XML comments in source code or supplemented with individual operator articles.
+* `Tutorials` - examples or tutorials about using the package. This section is optional, but valuable for more complex applications or when packages interoperate deeply with other packages.
```
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-style-guide.md:11]: Thanks, I will rephrase.

File: articles/documentation-docfx.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (banchan86) [reply articles/documentation-docfx.md:31]: Sounds good to me!

File: articles/documentation-docfx.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (banchan86) [reply articles/documentation-docfx.md:286]: Hmm I am not sure how to answer this, but I think there are two separate issues to consider here:

1) Unpublished vs published packages

I guess the distinction for me was writing documentation for new releases like Bonsai.PulsePal vs old releases like BonVision. 

It was a pain point for me when i started working on the Bonsai.PulsePal documentation because I initially did not know how to test the package out (which at that time was not released yet). I took what I learnt from the 'Creating a Package' guide and decided to write up this short section in case it helped other documentation contributors who were not package developers. However I do think that maybe it is a little out of place here and package developers writing documentation for their own packages would already have this knowledge. In that case maybe this section is unnecessary?

2) Bonsai workflow examples

I think consolidating the instructions for how to generate Bonsai example workflows might be a good idea as I do think its part of the core documentation infrastruture. I guess I was just struggling with how to organise this content and chose to document the import of the docfx-tools submodule as part of the docfx infrastructure and the saving of the Bonsai examples workflows as part of the media/content recommendations.

In that case, I would probably make a new separate section in this article and title it "Bonsai workflow support" and pull in the content about the importing of the docfx tools submodule, where to save the Bonsai workflows, and how to use the 'build.ps1' script. That leaves the markdown instructions for how to include them in articles. It might still be a good idea to leave in the style guide, along with a link to this "Bonsai workflow support" section to ensure that people set up the infrastructure correctly. If thats ok with you, I could do a rewrite in a new commit (unless you want to organise it yourself differently?)

File: articles/documentation-docfx.md (hunk 22 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md]: I ended up removing this bit.

File: articles/documentation-docfx.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (glopesdev) [reply articles/documentation-docfx.md:286]: Agreed, I have refactored both sections and moved the core infrastructure bit into the core docfx article.

**Review by glopesdev — APPROVED** [review]
